Implement WebView navigation events

jsuarezruiz opened the pull request:
### Description of Change ###

Implement WebView navigation events:
- Navigating
- Navigated

Fixes https://github.com/dotnet/maui/issues/3440

### PR Checklist ###

- [x] Targets the correct branch 
- [ ] Tests are passing (or failures are unrelated)

#### Does this PR touch anything that might affect accessibility?
- No

jsuarezruiz committed:
Created WebView navigating events

jsuarezruiz committed:
Implement events in Windows

jsuarezruiz committed:
Implemented on iOS

jsuarezruiz committed:
Implement events on Android

jsuarezruiz committed:
Added sample

jsuarezruiz committed:
Updated WebView Stub

jsuarezruiz committed:
Implement option to cancel navigation

Redth committed:
Merge branch 'main' into webview-navigation-events

jsuarezruiz committed:
Fix build error

jsuarezruiz committed:
Fix build error

PureWeen commented on src/Core/src/Handlers/WebView/WebViewHandler.Android.cs:
```diff
@@ -56,16 +68,23 @@ public static void MapWebViewSettings(WebViewHandler handler, IWebView webView)
 		public static void MapGoBack(WebViewHandler handler, IWebView webView, object? arg)
 		{
+			if (handler.NativeView.CanGoBack())
+				handler.CurrentWebNavigationEvent = WebNavigationEvent.Back;
```
The event state appears to get stuck with whatever it was set to.

- go to gallery for web view
- navigate forward
- click back
- now click any links on the page

You'll get a navigating event with a CurrentWebNavigationEvent set to `Back`

Maybe this call can just be centralized into the handler?

![image](https://user-images.githubusercontent.com/5375137/153514318-b0b0434b-b0d4-4a8d-83a5-7c6fc9153764.png)

jsuarezruiz replied:
Found the issue, this is the pending piece of code: https://github.com/xamarin/Xamarin.Forms/blob/caab66bcf9614aca0c0805d560a34e176d196e17/Xamarin.Forms.Platform.Android/Renderers/FormsWebViewClient.cs#L66

jsuarezruiz committed:
Merge branch 'main' into webview-navigation-events

jsuarezruiz committed:
Fix build errors

jsuarezruiz committed:
Updated handler

rmarinho reviewed (dismissed):
Seems something is making it crash on Android.. but not much info on the output.

Steps to reproduce:

- Go to WebView page
- Click dotnet
- Click GoBack
- Click GoForward
- Click GitHub logo
Crash

On windows when i click back it navigates again to the current page:

Navigating - Url: https://github.com/dotnet/maui, Event: Back
Navigated - Url: https://github.com/dotnet/maui, Event: Back, Result: Success
Navigating - Url: https://github.com/dotnet, Event: NewPage
Navigated - Url: https://github.com/dotnet, Event: NewPage, Result: Success

Redth committed:
Merge branch 'main' into webview-navigation-events

Redth committed:
Rebase on main, fix windows events

Now using the CoreWebView2 events on Windows which lets us use the history changed event so that updating cangoback/forward status works when navigating in the view.

Redth commented:
> Seems something is making it crash on Android.. but not much info on the output.
> 
> Steps to reproduce:
> 
> * Go to WebView page
> * Click dotnet
> * Click GoBack
> * Click GoForward
> * Click GitHub logo
>   Crash
> 
> On windows when i click back it navigates again to the current page:
> 
> Navigating - Url: https://github.com/dotnet/maui, Event: Back Navigated - Url: https://github.com/dotnet/maui, Event: Back, Result: Success Navigating - Url: https://github.com/dotnet, Event: NewPage Navigated - Url: https://github.com/dotnet, Event: NewPage, Result: Success

these are fixed with my changes

Redth committed:
Switch to interface for protocol

The apple docs state that if the newer decidePolicy method is implemented, the older one will not be called.  The concrete delegate implementation implements the newer one, so we can't use it without using the newer method which is only available on iOS13+.  To make it easier to still work on older versions, we can just implement the interface and export what we need instead.
 https://developer.apple.com/documentation/webkit/wknavigationdelegate/1455641-webview?language=objc#discussion

Redth commented:
Approving but created https://github.com/dotnet/maui/issues/4938 to track some minor issues on iOS/MacCatalyst 

/cc @mandel-macaque
